### PR TITLE
Airbyte CDK: move from gradle to poetry

### PR DESCRIPTION
**Pull request by erohmensing**
## What 
* Migrate from Gradle to Poetry for airbyte-cdk build
* Remove all gradle-related files 
* Remove java installation from CI pipelines and replace with installing poetry 
* Move setup.py implementation to poetry
* Move high-usage scrips to poetry `poe` tasks: 
  * Testing, linting, building (generator scripts) 
  * There is a separate PR to handle the mypy implementation since it has its own CI workflow and is more complicated with the changed files 
* Remove unused scripts: 
  * `low-code-unit-test.sh`
  * `validate-yaml-schema.sh`
* Update docs to stop referencing gradle: ⚠️ I did my best here. A lot of these docs are severely out of date and deserve a separate thorough lookover. TODO: create an issue enumerating docs that should be re looked over

### CODE REVIEW

**Comment by erohmensing** [comment]
* **#36250** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/36250?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @erohmensing and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by erohmensing** [comment]
Not quite ready, marking RFR for tests to run

File: airbyte-cdk/python/pyproject.toml [hunk 1]
```diff
@@ -1,44 +1,107 @@
...
+warn_return_any = true
+follow_imports = "silent"
+no_implicit_reexport = false
+#plugins = ["pydantic.mypy", "pendulum", "pytest-mypy-plugins"]
+
+[[tool.mypy.overrides]]
+module = "airbyte_cdk.models"
+ignore_errors = true
+
+[tool.poe.tasks]
```
Comment (alafanechere) [inline airbyte-cdk/python/pyproject.toml:98]: I'd love to have high level poe tasks exposed like:
* `poe lint` which would run flake
* `poe build` which would run the code gen + `poetry build` 
* `poe type_check` which would run mypy
* `poe test` which would run all the test suites of this project

I also found handy to have a `poe pre-push` command to use locally which would run all the tasks listed above.

**Review by alafanechere — APPROVED** [review]
Approving as nothing struck me as utterly wrong and I believe this 🧹 should be done incrementally otherwise the PR will get unreadable. So feel free to merge once you get 🟢 CI.

I'm not sure I get how you replaced `low-code-unit-tests` and `validate-yaml-schema` scripts.

IMO ideal follow up PRs would:
* Totally get rid of bash scripts
* Make the `build` poe task produce all the artifacts required for a pypi publication.
* Document the `build` poe task

In any case I'm convinced that this effort can help us write a dagger pipeline to publish the CDK.

File: .github/workflows/publish-cdk-command-manually.yml [hunk 2]
```diff
@@ -60,20 +60,25 @@ jobs:
...
+        uses: snok/install-poetry@v1
       - name: Checkout Airbyte
         uses: actions/checkout@v3
         with:
           repository: ${{ github.event.inputs.repo }}
           ref: ${{ github.event.inputs.gitref }}
+      - name: Install Dependencies
+        id: install_dependencies
+        working-directory: airbyte-cdk/python
+        run: poetry install --with dev
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply .github/workflows/publish-cdk-command-manually.yml:78]: Yep, you're right! I just copied this from another workflow, I wasn't sure if we were trying to be more explicit for users who are less familiar with poetry

File: airbyte-cdk/python/pyproject.toml [hunk 3]
```diff
@@ -1,44 +1,107 @@
...
+build-code-generator-image = "bin/build_code_generator_image.sh"
+generate-component-manifest-files = "bin/generate-component-manifest-files.sh"
+build = ["build-code-generator-image", "generate-component-manifest-files"]
+
+flake-check = "pflake8 --config ../../pyproject.toml ./"
+mypy-changed-files.cmd = "bin/run-mypy-on-modified-files.sh ${files}"
+mypy-changed-files.args = [{ name = "files", positional = true, multiple = true }]
+unit-test-with-cov = "coverage run --data-file=unit_tests/.coverage.testPython --rcfile=../../pyproject.toml -m pytest -s unit_tests -c pytest.ini"
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/pyproject.toml:84]: Pulled this straight from the [gradle file](https://github.com/airbytehq/airbyte/pull/36250/files#diff-25d60c0116afb59ec6c764847ffb9793fc4f20a02d17cad0848ac8a13dcf5149L124-L126) - Will try this option and see if the output looks the same! Seems way more straightforward to me

[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/pyproject.toml]: Switched, as far as I can tell we don't do anything with the coverage file at the moment, it's not even in a useful format. Made it terminal output for now - later we should probably provide a threshold

File: .github/workflows/python_cdk_tests.yml [hunk 4]
```diff
@@ -43,16 +43,28 @@ jobs:
...
-      - name: Install Pyenv
-        run: python3 -m pip install virtualenv --user
-      - name: Run Gradle Check on Python CDK
-        run: (cd airbyte-cdk/python && CI=true ./gradlew --no-daemon --no-build-cache --scan check)
+          path: "airbyte-cdk/python"
+          files: "airbyte_cdk/**/*.py"
+          diff_relative: true
+      - name: Run MyPy on modified files
+        working-directory: airbyte-cdk/python
+        run: poetry run poe mypy-changed-files ${{ steps.changed-py-files.outputs.all_changed_files }}
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply .github/workflows/python_cdk_tests.yml]: It's not even close to passing 😅 I agree that I don't like maintaining this changed files stuff (we calculate the changed files differently in ci than remote, too) 

I personally would prefer to have a big fix all the errors one error at a time by ignoring all of them, making a PR to fix all the `no-untyped-def`, another to fix the `no-untyped-call` etc. 

In trying to move the mypy stuff it actually got a bit rabbit holey so i pulled it out of this PR in any case

[the remark this answers is not in the document]
Comment (girarda) [reply .github/workflows/python_cdk_tests.yml]: > I personally would prefer to have a big fix all the errors one error at a time by ignoring all of them, making a PR to fix all the no-untyped-def, another to fix the no-untyped-call etc.
😨 
That would be great but I feel sorry for whoever would have to tackle this.

The changed files approach is hacky but I haven't found a better way to incrementally force the developers to fix the situation.

File: airbyte-cdk/python/pyproject.toml [hunk 5]
```diff
@@ -1,44 +1,107 @@
...
+    "Development Status :: 3 - Alpha",
+    "Intended Audience :: Developers",
+    "Topic :: Scientific/Engineering",
+    "Topic :: Software Development :: Libraries :: Python Modules",
+    "License :: OSI Approved :: MIT License",
+    "Programming Language :: Python :: 3.8",
 ]
+keywords = ["airbyte", "connector-development-kit", "cdk"]
 
-[tool.flake8]
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/pyproject.toml:23]: Yep! They are exactly the same 😄

File: airbyte-cdk/python/pyproject.toml [hunk 6]
```diff
@@ -1,44 +1,107 @@
...
+pytest-httpserver = "*"
+pytest-mock = "*"
+requests-mock = "*"
+Sphinx = "~4.2"
+sphinx-rtd-theme = "~1.0"
+tiktoken = "0.4.0"
+unstructured = { version = "0.10.27", extras = ["docx", "pptx"] }
+"unstructured.pytesseract" = ">=0.3.12"
+
+[tool.poetry.extras]
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/pyproject.toml:73]: This looks like the standard way to format the extras according to [poetry docs](https://python-poetry.org/docs/pyproject/#extras)! I'll keep it as is

File: .github/workflows/publish-cdk-command-manually.yml [hunk 7]
```diff
@@ -60,20 +60,25 @@ jobs:
...
-      - uses: actions/setup-python@v4
+      - name: Install Python
+        uses: actions/setup-python@v4
         with:
           python-version: "3.10"
-      - uses: actions/setup-java@v3
-        with:
-          distribution: "zulu"
-          java-version: "21"
+      - name: Install Poetry
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply .github/workflows/publish-cdk-command-manually.yml:67]: At the moment, we don't have any github actions that install poetry another way. I think your reasoning is valid and we should make sure we're considering this when we build + release the CDK via airbyte CI!

File: airbyte-cdk/python/pyproject.toml [hunk 8]
```diff
@@ -1,44 +1,107 @@
...
+pytest = "6.2.5"
+pytest-cov = "*"
+pytest-httpserver = "*"
+pytest-mock = "*"
+requests-mock = "*"
+Sphinx = "~4.2"
+sphinx-rtd-theme = "~1.0"
+tiktoken = "0.4.0"
+unstructured = { version = "0.10.27", extras = ["docx", "pptx"] }
+"unstructured.pytesseract" = ">=0.3.12"
```
[the remark this answers is not in the document]
Comment (erohmensing) [reply airbyte-cdk/python/pyproject.toml]: I _think_ it's a way to define the version in one place and then reuse it in extras. Maybe this is stranger due to the fact that poetry includes dev dependencies by default whereas setup.py does not (therefore defining the versions in the "give me everything" extra and reusing those in the subset extras makes more sense) 

I'll see if there is an easy win here, but yeah I would feel more comfortable if those who defined these extras would be willing to look into it to split things up

File: airbyte-cdk/python/pyproject.toml (hunk 1 above)
Comment (erohmensing) [reply airbyte-cdk/python/pyproject.toml:78]: Good ideas! Updated to add some more, now we have 
```
CONFIGURED TASKS
  build-code-generator-image
  generate-component-manifest-files
  generate
  build-package
  build
  flake-check
  lint
  type-check
  unit-test-with-cov
  check
  pre-push
  ```

**Comment by erohmensing** [comment]
@alafanechere 

> I'm not sure I get how you replaced low-code-unit-tests and validate-yaml-schema scripts.

I 🪓 them with Brian Lai's permission! Better Pr description should help here. 

```
Totally get rid of bash scripts
Make the build poe task produce all the artifacts required for a pypi publication.
Document the build poe task
```
Agreed. I think our current publish via github action doesn't need the artifacts (well, i'm about to find out!) but as we move to airbyte ci, this makes sense. In the meantime before fleshed out docs, I added help info to all of the `poe` tasks.

**Comment by octavia-approvington** [comment]
A job well done <br/> nice job

**Review by octavia-approvington — APPROVED** [review]

**Pull request merged**